### PR TITLE
Fix FileWatcher

### PR DESCRIPTION
**Pull request by CookiePLMonster**
This PR fixes several issues with FileWatcher, some of them Windows-only:

Windows-only:
* Fixed Unicode path issues.
* Fixed file sharing issues (the watcher handle actually prevented modifications of the watched directory, making the entire ordeal rather pointless).
* Guarded the code against `ReadDirectoryChangesW` returning 0 bytes while also returning `true`, as the docs specifically state it's possible.

General:
* Simplified the watcher thread creation semantics and made it join **before** closing watcher handles, just to be sure.
* Made `ScriptEngine::AutoReloadPlugins()` take a mutex earlier as checking if the container is not empty while not holding a mutex is probably a bad idea too.
* More `string_view` goodness

It's nitpicks so I don't think they're worth mentioning in the changelog, the only user-observable issue was the sharing issue anyway.

### CODE REVIEW

**Comment by Broxzier** [comment]
This partly supersedes https://github.com/OpenRCT2/OpenRCT2/pull/17833

File: src/openrct2/scripting/ScriptEngine.cpp
Comment (IntelOrca) [inline src/openrct2/scripting/ScriptEngine.cpp:821]: What is the reason behind this change?

Comment (CookiePLMonster) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: > Made ScriptEngine::AutoReloadPlugins() take a mutex earlier as checking if the container is not empty while not holding a mutex is probably a bad idea too.

This is a diff from moving everything a single indent to the left by removing `if (_changedPluginFiles.size() > 0)`, as a result all operations on the vector are under a mutex.

Comment (IntelOrca) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: The reason it checked if the vector was not empty first was to improve performance. It would only lock if needed to enumerate the vector. Reading the size of the vector should be a thread safe op.

Comment (Broxzier) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: It may change between reading the length and accessing it, so even if reading the length of the vector is atomic, it's not thread safe.

Comment (CookiePLMonster) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: I don't think it should be a concern here - this check is done only once per frame, and mutexes are very fast to lock when not under contention. Even if it's a thread safe operation, you risk unneccessary race conditions, albeit harmless ones.

Comment (IntelOrca) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: > It may change between reading the length and accessing it, so even if reading the length of the vector is atomic, it's not thread safe.

That is what I meant, if it is atomic that is fine. It just means that it won't process the loop until next time round. This is to reduce frequeny of a lock.

Comment (CookiePLMonster) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: It's still negligible, one uncontented lock per frame is nothing.

Comment (IntelOrca) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: That was my reason for putting it there in the first place, is there a good reason for removing other than "it is negligible"? It is +1 lock per frame.

Comment (CookiePLMonster) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: > is there a good reason for removing other than "it is negligible"?

Correctness

Comment (Broxzier) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: I agree with CookiePLMonster that it's better to avoid any data racing than to assume that the next frame the size will match.

Comment (IntelOrca) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: Okay, I personally disagree, I have always implemented this concept for both personal projects, OpenRCT2 and work.

Comment (CookiePLMonster) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: I can skip this entire function, and thus the mutex, if `_plugins.empty()` since if that is the case then the for loop will never do any action anyway - is that a reasonable compromise?

Comment (IntelOrca) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: I would have no problem approving it if it were just the changes for the fix. The string_view is fine as well, although @Broxzier did point out the overlap with another PR.

Comment (Broxzier) [reply src/openrct2/scripting/ScriptEngine.cpp:821]: Don't mind the other PR, I think that one should be made smaller anyway.

**Review by Broxzier — APPROVED** [review]

File: src/openrct2/scripting/ScriptEngine.cpp
Comment (CookiePLMonster) [reply src/openrct2/scripting/ScriptEngine.cpp]: OK, I reintroduced the emptiness check before taking the mutex so this PR doesn't stay unmerged.

**Review by IntelOrca — APPROVED** [review]

**Comment by tupaschoal** [comment]
Given this has 2 approvals, I'm merging this.

**Pull request merged**
